core/src/main/kotlin/in/specmatic/core/Resolver.kt

```diff
@@ -69,9 +70,35 @@ data class Resolver(
             else -> throw ContractException("$patternValue is not a type")
         }
 
+    private fun nullable(orig: Pattern): Boolean {
+        val pattern = if (orig is DeferredPattern) orig.resolvePattern(this) else orig
+        return pattern is AnyPattern && pattern.pattern.contains(NullPattern)
+    }
+
+    private fun validateNoCycle(pattern: Pattern) {
+        val index = cyclePreventionStack.indexOf(pattern)
+        if (index >= 0) {
+            // Nullables are allowed since they will eventually terminate any cycle (only consider substack with cycle)
+            val subStack = cyclePreventionStack.subList(index, cyclePreventionStack.size)
+            if (subStack.stream().noneMatch { nullable(it) }) {
+                // Terminate what would otherwise be an infinite cycle.
+                throw PatternCycleException("Invalid pattern cycle", cyclePreventionStack.plus(pattern))
+            }
+        }
+    }
+
     fun generate(factKey: String, pattern: Pattern): Value {
-        if (!factStore.has(factKey))
-            return pattern.generate(this)
+        if (!factStore.has(factKey)) {
+
+            validateNoCycle(pattern)
+
+            cyclePreventionStack.add(pattern)
```

Comment:
@westse We notice that cyclePreventionStack is a multable list which is an instance variable of Resolver, which thereby makes Resolver itself mutable. Because of the nature of how we pass resolver around, it is extremely critical to keep the resolve immutable. We will make the necessary changes (passing the type stack as method parameter thereby not requiring instance variable) and push it your branch. cc @westse